File: solutions/python/binary-search/search-a-2d-matrix-ii.py
# LeetCode 240 — Search a 2D Matrix II (Medium)
# Category: Binary Search · Approach: Staircase
# Time: O(m + n) | Space: O(1)
# Source: https://leetcode.com/problems/search-a-2d-matrix-ii/

def search_matrix(matrix: list[list[int]], target: int) -> bool:
    if not matrix or not matrix[0]:
        return False
    r, c = 0, len(matrix[0]) - 1
    while r < len(matrix) and c >= 0:
        v = matrix[r][c]
        if v == target:
            return True
        if v > target:
            c -= 1
        else:
            r += 1
    return False

File: solutions/typescript/binary-search/search-a-2d-matrix-ii.ts
// LeetCode 240 — Search a 2D Matrix II (Medium)
// Category: Binary Search · Approach: Staircase
// Time: O(m + n) | Space: O(1)
// Source: https://leetcode.com/problems/search-a-2d-matrix-ii/

function searchMatrix(matrix: number[][], target: number): boolean {
  if (!matrix.length || !matrix[0].length) return false;
  let r = 0, c = matrix[0].length - 1;
  while (r < matrix.length && c >= 0) {
    const v = matrix[r][c];
    if (v === target) return true;
    if (v > target) c--;
    else r++;
  }
  return false;
}
